rust/trace/src/lib.rs

```diff
@@ -393,18 +482,22 @@ impl<'a> SampleGuard<'a> {
         -> Self
         where S: Into<StrCow>, C: Into<CategoriesT>
     {
-        Self {
-            sample: Some(Sample::new(name, categories, payload)),
+        let guard = Self {
+            sample: Some(Sample::new_duration_marker(
+                name, categories, payload, SampleEventType::DurationBegin)),
             trace: Some(&trace),
-        }
+        };
+        trace.record(&guard.sample.as_ref().unwrap());
+        guard
     }
 }
 
 impl<'a> Drop for SampleGuard<'a> {
     fn drop(&mut self) {
-        if let Some(ref mut sample) = self.sample {
-            sample.set_end_ns(time::precise_time_ns());
-            self.trace.unwrap().record(sample);
+        if let Some(ref mut trace) = self.trace {
+            let sample = self.sample.as_mut().unwrap();
+            sample.event_type = SampleEventType::DurationEnd;
```

Comment:
Don't we still need to set the end time here? It's confusing to me that we treat guard blocks differently than we treat closure blocks.

Reply:
There's no more end time.  Duration is different from start/stop in the Chrome Trace format.  There's no good clarification in the docs on where you would use either, but using the duration type for closures should be slightly more efficient (cpu & memory) since it only creates 1 Sample entry & inserts it once.

Reply:
In that case I think we should use the duration type for blocks as well? Certainly the way I've been using them I expect them to show up as spans in the trace viewer. There are a lot of situations where `trace_block` is more ergonomic than `trace_closure`, but conceptually I think they perform the same task.

In either case I'm confused by what's going on in this bit of code; my understanding is that we're adding a new sample, that has a distinct `SampleEventType` but otherwise is the same as the first. They even have the same `sample_id`?

Reply:
Also following this back to `Trace::record`, and noticed that we're always cloning the sample. I think instead `Trace::record` should be taking ownership of the sample, and if the caller needs to retain a copy they should `clone()` explicitly. In the general case this shouldn't be necessary, so it's currently an extra clone on every trace.

Reply:
It's a little less than straightforward (i.e. i'm likely to accidentally insert a bug) to have the block sample path coalesce the sample into a duration event so my instinct is to punt on it for now & leave a TODO.  In the UI I don't believe there's a difference between trace_block & trace_closure so the only difference should be the slightly larger JSON depending on how much more frequently trace_block is used.

I've corrected the perf issue you noted with Trace::record.